huff0/decompress_amd64.s

```diff
@@ -0,0 +1,337 @@
+#include "textflag.h"
+#include "funcdata.h"
+#include "go_asm.h"
+
+#define mask        $0x7ff  // up to 11 bits, according to the spec
+#define bufoff      256     // see decompress.go, we're using [4][256]byte
+
+//func decompress4x_main_loop_bmi1(pbr0, pbr1, pbr2, pbr3 *bitReaderShifted,
+//	peekBits uint8, buf *byte, tbl *dEntrySingle) (int, bool)
+TEXT ·decompress4x_main_loop_bmi1(SB), NOSPLIT, $0
+#define off             R8
+#define buffer          DI
+#define table           SI
+
+#define br_bits_read    R9
+#define br_value        R10
+#define br_offset       R11
+#define peek_bits       R12
+#define exhausted       DX
+
+#define br0             R13
+#define br1             R14
+#define br2             R15
+#define br3             BP
+
+    XORQ    exhausted, exhausted    // exhausted = false
+    XORQ    off, off                // off = 0
+
+    MOVBQZX peekBits+32(FP), peek_bits
+    MOVQ    buf+40(FP), buffer
+    MOVQ    tbl+48(FP), table
+
+    MOVQ    pbr0+0(FP), br0
+    MOVQ    pbr1+8(FP), br1
+    MOVQ    pbr2+16(FP), br2
+    MOVQ    pbr3+24(FP), br3
+
+main_loop:
+
+    // const stream = 0
+    // br0.fillFast()
+    MOVBQZX bitReaderShifted_bitsRead(br0), br_bits_read
+    MOVQ    bitReaderShifted_value(br0), br_value
+    MOVQ    bitReaderShifted_off(br0), br_offset
+
+	// if b.bitsRead >= 32 {
+    CMPQ    br_bits_read, $32
+    JB      skip_fill0
+
+    SUBQ    $32, br_bits_read       // b.bitsRead -= 32
+    SUBQ    $4, br_offset           // b.off -= 4
+
+	// v := b.in[b.off-4 : b.off]
+	// v = v[:4]
+	// low := (uint32(v[0])) | (uint32(v[1]) << 8) | (uint32(v[2]) << 16) | (uint32(v[3]) << 24)
+    MOVQ    bitReaderShifted_in(br0), AX
+    MOVL    0(br_offset)(AX*1), AX  // AX = uint32(b.in[b.off:b.off+4])
+
+	// b.value |= uint64(low) << (b.bitsRead & 63)
+    SHLXQ   br_bits_read, AX, AX
+    ORQ     AX, br_value
+    // }
+skip_fill0:
+
+    // exhausted = exhausted || (br0.off < 4)
+    CMPQ    br_offset, $4
+    SETLT   DL
+    ORB     DL, DH
+
+    // val0 := br0.peekTopBits(peekBits)
+    SHRXQ   peek_bits, br_value, AX
+    ANDQ    mask, AX                // AX = (value >> peek_bits) & mask
+
+    // v0 := table[val0&mask]
+    MOVW    0(table)(AX*2), AX      // AX - v0
+
+    // br0.advance(uint8(v0.entry))
+    MOVBQZX AL, CX
+    SHLXQ   CX, br_value, br_value  // value <<= n
+    ADDQ    CX, br_bits_read        // bits_read += n
+
+    MOVB    AH, BL                  // BL = uint8(v0.entry >> 8)
+
+    // val1 := br0.peekTopBits(peekBits)
+    SHRXQ   peek_bits, br_value, AX
+    ANDQ    mask, AX                // AX = (value >> peek_bits) & mask
```

Comment:
Not needed.  Just in Go code for bounds check elimination.

Reply:
Didn't know that, thanks.